clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -289,6 +314,32 @@ public synchronized void failIfNotReadyForSend() {
         }
     }
 
+    public void awaitResultOrThrowTimeoutException(long maxBlockTimeMs,
```

Comment:
As an alternative, we could move this logic into `TransactionalRequestResult`. We can change `TransactionalRequestResult.await(timeout)` to raise a suitable timeout and handle the `InterruptedException`. We can also pass the operation as a string when constructing `TransactionalRequestResult`.  For example:
```java
    public boolean await(long timeout, TimeUnit unit) {
        try {
            boolean completed = latch.await(timeout, unit);
            if (!completed)
                throw new TimeoutException("Timeout expired after" + timeout + "ms while awaiting " + operation);

            if (!isSuccessful())
                throw error();

            return true;
        } catch (InterruptedException e) {
            throw new InterruptException("Received interrupt while awaiting " + operation, e);
        }
    }
```